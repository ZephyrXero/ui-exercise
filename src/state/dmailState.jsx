// I've opted for a simple state manager via React's Context system
import React, { useEffect } from 'react';

const SERVER = "http://localhost:3000"; // TODO: choose dynamically from environment
const initialState = { currentEmail: null };

export const DMailState = React.createContext();

export function DMailPageState(props) {
  const { children } = props;
  const [state, setState] = React.useState(initialState);

  useEffect(() => {
    getEmails();
  }, []);

  return <DMailState.Provider value={state}>{children}</DMailState.Provider>

  async function getEmails() {
    const FILE_PATH = "/emails.json";
    const emails = await fetch(SERVER + FILE_PATH)
      .then(response => response.json())
      .catch(error => console.error('Data file failed to load'));

    setState(priorState => {
      return { ...priorState, ...{ emailList: emails }};
    });
  }
}

export default DMailState;
