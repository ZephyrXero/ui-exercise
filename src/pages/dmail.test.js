import React from 'react';
import { render } from '@testing-library/react';

import DMail from './dmail';
import { DMailState } from '../state/dmailState';

describe('DMail Page', () => {
  const testState = {
    currentEmail: null,
    emailList: [
      {
        id: '1',
        subject: 'Hello',
        sender: 'bob.smith@gmail.com',
        body: 'test body',
        tags: ['work'],
        date: '2017-03-05T03:25:43.511Z',
      },
      {
        id: '2',
        subject: 'Important information about your account',
        sender: 'jon.doe@gmail.com',
        body: 'test message about account',
        tags: ['travel'],
        date: '2020-03-27T10:25:43.511Z',
      },
    ]
  };

  test('should render with list', () => {
    const page = render(
      <DMailState.Provider value={testState}>
        <DMail />
      </DMailState.Provider>
    );
    const list = page.getByTestId('email-list');

    expect(page.container.firstChild).toMatchSnapshot();
    expect(list).toBeInTheDocument();
  });

  test.skip('should show individual emails when clicked', () => {
    // make sure test email displays once clicked

    // expect(testSubject).toBe(expectedSubject)
    // expect(lastWordInBody).toBe(expectedWord)
  });
});
