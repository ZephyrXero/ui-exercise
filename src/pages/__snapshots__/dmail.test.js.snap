// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DMail Page should render with list 1`] = `
<div
  id="dmail"
>
  <div
    id="navbar"
  >
    <div
      id="menu-column"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
        />
      </svg>
      <span
        class="logo-text"
      >
        DMail
      </span>
    </div>
    <div
      id="search-column"
    >
      <div
        id="search-bar"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
        Search Mail
      </div>
    </div>
    <div
      id="user-column"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
        />
      </svg>
    </div>
  </div>
  <div
    class="flex-row"
  >
    <div
      id="sidebar"
    >
      <button
        id="compose-button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
         Compose
      </button>
      <button
        class="sidebar-item flat-button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H4.99c-1.11 0-1.98.89-1.98 2L3 19c0 1.1.88 2 1.99 2H19c1.1 0 2-.9 2-2V5c0-1.11-.9-2-2-2zm0 12h-4c0 1.66-1.35 3-3 3s-3-1.34-3-3H4.99V5H19v10z"
          />
        </svg>
         Inbox
      </button>
      <span
        class="sidebar-item"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
         Starred
      </span>
      <span
        class="sidebar-item"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
          />
        </svg>
         Sent
      </span>
      <span
        class="sidebar-item"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 2c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6H6zm7 7V3.5L18.5 9H13z"
          />
        </svg>
         Drafts
      </span>
      <span
        class="sidebar-item"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.73 3H8.27L3 8.27v7.46L8.27 21h7.46L21 15.73V8.27L15.73 3zM12 17.3c-.72 0-1.3-.58-1.3-1.3 0-.72.58-1.3 1.3-1.3.72 0 1.3.58 1.3 1.3 0 .72-.58 1.3-1.3 1.3zm1-4.3h-2V7h2v6z"
          />
        </svg>
         Spam
      </span>
      <button
        class="sidebar-item flat-button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17.63 5.84C17.27 5.33 16.67 5 16 5L5 5.01C3.9 5.01 3 5.9 3 7v10c0 1.1.9 1.99 2 1.99L16 19c.67 0 1.27-.33 1.63-.84L22 12l-4.37-6.16z"
          />
        </svg>
         
        work
      </button>
      <button
        class="sidebar-item flat-button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17.63 5.84C17.27 5.33 16.67 5 16 5L5 5.01C3.9 5.01 3 5.9 3 7v10c0 1.1.9 1.99 2 1.99L16 19c.67 0 1.27-.33 1.63-.84L22 12l-4.37-6.16z"
          />
        </svg>
         
        travel
      </button>
    </div>
    <div
      id="content-container"
    >
      <div
        data-testid="email-list"
        id="email-list"
      >
        <div
          id="list-header"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
            />
          </svg>
          <span
            class="group-label"
          >
            Unread
          </span>
          <div
            class="email-counter"
          >
            1-2
             of 
            2
          </div>
        </div>
        <div
          class="email-preview"
        >
          <span
            class="sender"
          >
            bob.smith@gmail.com
          </span>
          <span
            class="subject"
          >
            Hello
          </span>
          <span
            class="body"
          >
            test body
          </span>
          <span
            class="date"
          >
            3/4/2017, 10:25:43 PM
          </span>
        </div>
        <div
          class="email-preview"
        >
          <span
            class="sender"
          >
            jon.doe@gmail.com
          </span>
          <span
            class="subject"
          >
            Important information about your account
          </span>
          <span
            class="body"
          >
            test message about account
          </span>
          <span
            class="date"
          >
            3/27/2020, 6:25:43 AM
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
